### PR TITLE
[tune] [rllib] Allow checkpointing to object store instead of local disk

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Adds `save_to_object()` and `restore_from_object()`, which supplement `save()` and `restore()`. The advantage of these is that they use the Ray object store which simplifies file management. To avoid using too much space, we gzip compress checkpoints.

This also switches ray.tune to use object checkpoints by default, to support clustered mode.

### CODE REVIEW

**Comment by richardliaw** [comment]
@ericl a couple tests are failing

File: python/ray/rllib/agent.py [hunk 1]
```diff
@@ -195,6 +197,34 @@ def save(self):
...
         return checkpoint_path
 
+    def save_to_object(self):
+        """Saves the current model state to the Ray object store.
+
+        Returns:
+            ObjectID of the checkpoint.
+        """
+
+        checkpoint_prefix = self.save()
```
Comment (richardliaw) [inline python/ray/rllib/agent.py]: this is a write to disk; does it make sense to store this checkpoint prefix in the agent as a checkpoint path too (as a fallback)? We would reorder the restoring

File: python/ray/rllib/agent.py [hunk 2]
```diff
@@ -195,6 +197,34 @@ def save(self):
             open(checkpoint_path + ".rllib_metadata", "wb"))
         return checkpoint_path
 
+    def save_to_object(self):
+        """Saves the current model state to the Ray object store.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/agent.py]: Fixed to just drop the object store mention.

File: python/ray/rllib/agent.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/agent.py]: Writes to disk are fast sine in the common case you hit fs cache.

File: python/ray/rllib/agent.py [hunk 3]
```diff
@@ -208,6 +238,25 @@ def restore(self, checkpoint_path):
...
+
+        out = StringIO.StringIO(obj)
+        info = pickle.loads(gzip.GzipFile(fileobj=out).read())
+        data = info["data"]
+        tmpdir = tempfile.mkdtemp("restore_from_object", dir=self.logdir)
+        checkpoint_path = os.path.join(tmpdir, info["checkpoint_name"])
+
+        for file_name, file_contents in data.items():
+            with open(os.path.join(tmpdir, file_name), "wb") as f:
+                f.write(file_contents)
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/agent.py]: Tensorflow seems to expect to read from disk, therefore we have to go through this in any case

**Review by richardliaw — APPROVED** [review]

**Comment by ericl** [comment]
jenkins retest this please

**Comment by ericl** [comment]
Oh hm, I see the last jenkins result went back to FAILED. Though it seems to be some unrelated error: `java.io.IOException: Failed to mkdirs: /home/jenkins/workspace/Ray-PRB`, hopefully it was still ok.

**Comment by robertnishihara** [comment]
Most likely unrelated.

**Comment by JourneyGo** [comment]
I got error from running this line in save_to_object()
`            if path.startswith(checkpoint_prefix):
                data[os.path.basename(path)] = open(path, "rb").read()` 

The full function is:
`def save(self, checkpoint_dir=None):
        """Saves the current model state to a checkpoint.

        Subclasses should override ``_save()`` instead to save state.
        This method dumps additional metadata alongside the saved path.

        Args:
            checkpoint_dir (str): Optional dir to place the checkpoint.

        Returns:
            Checkpoint path that may be passed to restore().
        """

        checkpoint_path = self._save(checkpoint_dir or self.logdir)
        pickle.dump([
            self._experiment_id, self._iteration, self._timesteps_total,
            self._time_total, self._episodes_total
        ], open(checkpoint_path + ".tune_metadata", "wb"))
        return checkpoint_path

    def save_to_object(self):
        """Saves the current model state to a Python object. It also
        saves to disk but does not return the checkpoint path.

        Returns:
            Object holding checkpoint data.
        """

        tmpdir = tempfile.mkdtemp("save_to_object", dir=self.logdir)
        checkpoint_prefix = self.save(tmpdir)

        data = {}
        base_dir = os.path.dirname(checkpoint_prefix)
        for path in os.listdir(base_dir):
            path = os.path.join(base_dir, path)
            if path.startswith(checkpoint_prefix):
                data[os.path.basename(path)] = open(path, "rb").read()

        out = io.BytesIO()
        with gzip.GzipFile(fileobj=out, mode="wb") as f:
            compressed = pickle.dumps({
                "checkpoint_name": os.path.basename(checkpoint_prefix),
                "data": data,
            })
            if len(compressed) > 10e6:  # getting pretty large
                logger.info("Checkpoint size is {} bytes".format(
                    len(compressed)))
            f.write(compressed)

        shutil.rmtree(tmpdir)
        return out.getvalue()`

According to the code, save_to_object use save(tmpdir) at first. The save() function save .ckpt file into tmpdir and make a tmpdir.tune_metadata file into the tmpdir parent dir. But when save() return, save_to_object function will try to read all file(path) in the tmpdir's parent dir with which name startwith tmpdir. 
So the open('ckpt_path','rb').read() will return error.
Is there anything wrong above? Or this is a bug or something :(

**Comment by ericl** [comment]
Can you file an issue? This looks like a bug but needs more details to
reproduce.

+Richard

**Comment by JourneyGo** [comment]
Sure

**Pull request merged**